File: 0827-Making A Large Island/0827-Making A Large Island.java
class Solution {
    public int largestIsland(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[] parents = new int[m * n], areas = new int[m * n];
        for (int i = 0; i < m * n; ++i) {
            parents[i] = i;
        }

        int maxArea = 0;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 1) {
                    int p1 = i * n + j;
                    areas[p1] = 1;
                    if (i > 0 && grid[i - 1][j] == 1) {
                        p1 = findParent(parents, (i - 1) * n + j);
                        parents[i * n + j] = p1;
                        ++areas[p1];
                    }
                    if (j > 0 && grid[i][j - 1] == 1) {
                        p1 = findParent(parents, i * n + j - 1);
                        int p2 = findParent(parents, i * n + j);
                        if (p1 != p2) {
                            parents[p2] = p1;
                            areas[p1] += areas[p2]; 
                        }                           
                    }
                    maxArea = Math.max(maxArea, areas[p1]);
                }
            }
        }
        
        int[] dr = {-1, 1, 0, 0};
        int[] dc = {0, 0, -1, 1};
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 0) {
                    areas[i * n + j] = 1;
                    Set<Integer> visited = new HashSet();
                    for (int k = 0; k < 4; ++k) {
                        int r = i + dr[k];
                        int c = j + dc[k];
                        if (r >= 0 && r < m && c >= 0 && c < n && grid[r][c] == 1) {                            
                            int p = findParent(parents, r * n + c);
                            if (!visited.contains(p)) {
                                visited.add(p);
                                areas[i * n + j] += areas[p];
                            }                            
                        }                        
                    }
                    maxArea = Math.max(maxArea, areas[i * n + j]);
                }
            }
        }
        return maxArea;
    }
    
    private int findParent(int[] parents, int i) {
        while (parents[i] != i) {
            parents[i] = parents[parents[i]];
            i = parents[i];
        }
        return i;
    }
}

File: 0801-0900/0827-Making A Large Island/0827-Making A Large Island.py
class Solution:
    def largestIsland(self, grid: List[List[int]]) -> int:                
        m = len(grid)
        n = len(grid[0])
        parent = [i for i in range(m * n)]
        sz = [1] * (m*n)
        def find(i):
            while i != parent[i]:
                parent[i] = parent[parent[i]]
                i = parent[i]
            return i
        
        def union(p, q):
            rootP = find(p)
            rootQ = find(q)
            if rootP == rootQ:
                return
            if sz[rootP] < sz[rootQ]:
                parent[rootP] = rootQ
                sz[rootQ] += sz[rootP]
            else:
                parent[rootQ] = rootP
                sz[rootP] += sz[rootQ]
                
        def area(i):
            return sz[find(i)]
                
        maxArea = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    index = i * n + j
                    if i > 0 and grid[i - 1][j] == 1:
                        union(index, (i - 1)*n + j)                        
                    if j > 0 and grid[i][j - 1] == 1:
                        union(index, i*n + j - 1)
                    maxArea = max(maxArea, area(index))
                    
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 0:
                    curr = 1
                    visited = set()
                    for r, c in (i-1, j), (i+1, j), (i, j-1), (i, j+1):
                        if 0 <= r < m and 0 <= c < n and grid[r][c] == 1:
                            p = find(r * n + c)
                            if p not in visited:
                                visited.add(p)
                                curr += sz[p]
                    
                    maxArea = max(maxArea, curr)
        return maxArea
